src/Builder/BuilderEncoder.php

```diff
@@ -4,323 +4,101 @@
-                }
-
-                $result->{$key} = $this->encodeIfSupported($value);
-            }
+        if (! $encoder || ! $encoder->canEncode($value)) {
+            throw UnsupportedValueException::invalidEncodableValue($value);
         }
 
-        return $result;
+        return $encoder->encode($value);
     }
 
-    /**
-     * For the $setWindowFields stage output parameter, the optional window parameter is encoded in the same object
-     * of the window operator.
-     *
-     * @see https://www.mongodb.com/docs/manual/reference/operator/aggregation/setWindowFields/
-     */
-    private function encodeOutputWindow(OutputWindow $outputWindow): stdClass
+    private function getEncoderFor(object $value): ExpressionEncoder|null
     {
-        $result = $this->recursiveEncode($outputWindow->operator);
-
-        // Transform the result into an stdClass if a document is provided
-        if (! $outputWindow->operator instanceof WindowInterface && (is_array($result) || is_object($result))) {
-            if (! is_first_key_operator($result)) {
-                throw new LogicException(sprintf('Expected OutputWindow::$operator to be an operator. Got "%s"', array_key_first((array) $result)));
-            }
-
-            $result = (object) $result;
+        $valueClass = $value::class;
+        if (isset($this->cachedEncoders[$valueClass])) {
+            return $this->cachedEncoders[$valueClass];
         }
 
-        if (! $result instanceof stdClass) {
-            throw new LogicException(sprintf('Expected OutputWindow::$operator to be an stdClass, array or WindowInterface. Got "%s"', get_debug_type($result)));
-        }
+        $encoderList = $this->customEncoders + $this->defaultEncoders;
 
-        if ($outputWindow->window !== Optional::Undefined) {
-            $result->window = $this->recursiveEncode($outputWindow->window);
-        }
-
-        return $result;
-    }
-
-    /**
-     * Nested arrays and objects must be encoded recursively.
-     */
-    private function recursiveEncode(mixed $value): mixed
-    {
-        if (is_array($value)) {
-            foreach ($value as $key => $val) {
-                $value[$key] = $this->recursiveEncode($val);
+        // First attempt: match class name exactly
+        foreach ($encoderList as $className => $encoderClass) {
+            if ($className == $valueClass) {
+                return $this->cachedEncoders[$valueClass] = new $encoderClass($this);
             }
-
-            return $value;
         }
 
-        if ($value instanceof stdClass) {
-            foreach (get_object_vars($value) as $key => $val) {
-                $value->{$key} = $this->recursiveEncode($val);
+        // Second attempt: catch child classes
+        foreach ($encoderList as $className => $encoderClass) {
+            if ($value instanceof $className) {
+                return $this->cachedEncoders[$valueClass] = new $encoderClass($this);
```

Comment:
This will construct encoder instances for each subclass. Do you want to leave a comment about potentially restructuring this to be more efficient?

Likewise, you can also optimize for the miss case. It's not as important for `encode()` since that's an error condition but would help with `canEncode()` calls if those are more frequent.

Maybe this is something to ticket for potential profiling and refactoring at a later point.

Reply:
Forgot to mention this: I refactored the logic:
a) `null` values for classes with no encoders are now cached
b) The "first" attempt section now uses `isset` instead of looping through the list to check the key (🤦‍♂️)